text/0000-event-handler.md

```diff
@@ -0,0 +1,122 @@
+- Feature Name: EventHandler
+- Start Date: 2018-11-30
+- RFC PR:
+- Neon Issue:
+
+# Summary
+[summary]: #summary
+
+The goal of this RFC is to provide a way for bindings to call back into JavaScript from threads other than the Node.JS main thread.
+
+# Motivation
+[motivation]: #motivation
+
+The main motivation of this is to provide a way to schedule the execution of JavaScript from any (Rust) thread. While `neon::Task` allows to perform a background task and call a JavaScript callback after it's completion, there is currently no way to propagate the progress of a task to JavaScript.   
+Being able to call to JavaScript from any thread would be useful for binding authors trying to bind native libraries:
+- with custom threading models
+- with progress state
+- which produce events
+- with callbacks
+- ...
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC introduces a new struct `neon::event::EventHandler` that is clone-able and can be send across threads.    
+The user provides a `function`  (and an optional `this` context which defaults to the global object) to the struct constructor.
+To actually invoke JavaScript code the `schedule` method is used. This method accepts a closure which will receive a `neon::Context` and the `this`/`function` pair provided during the creation of `neon::event::EventHandler`.
+
+Example for providing the current progress of a background operation:
+
+```rust
+    let mut this = cx.this();
+    let func = cx.argument::<JsFunction>(0)?;
+    let cb = EventHandler::new(this, func);
+    thread::spawn(move || {
+        for i in 0..100 {
+            // do some work ....
+            thread::sleep(Duration::from_millis(40));
+            // call into javascript
+            cb.schedule(move |context, this, func| {
+                // call the js function
+                func.call(context, this, vec![cx.number(i).upcast()]);
+            }
+        }
+    });
+```
+
+Here the `EventHandler` "captures" `this` and `func` and calls the closure from the JavaScript thread with a `context`, the `this` and `func` value.
+
+*Note:* The closure is send to the main Javascript thread so every captured value will be moved.   
+
+This approach should be familiar to every Rust programmer as it is the same as `std::thread::spawn` uses.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The struct is implemented as following:
+
+```rust
+struct EventHandlerInner(*mut c_void);
+
+unsafe impl Send for EventHandlerInner {}
+unsafe impl Sync for EventHandlerInner {}
+
+impl Drop for EventHandlerInner {
+    fn drop(&mut self) {
+        // free resources
+    }
+}
+
+#[derive(Clone)]
+pub struct EventHandler(Arc<EventHandlerInner>);
+
+impl EventHandler {
+    pub fn new<T: Value>(this: Option<Handle<T>>, function: Handle<JsFunction>) -> Self;
```

Comment:
I made `this` an `Option` for now, not sure if we want to have two constructor functions one which takes `this` and `function` and one which only takes a `function` as argument....

Reply:
I like the option.

Reply:
As discussed on slack I changed this to `EventHandler::new` and `EventHandler::bind`